### PR TITLE
`Learning path`: Redesign learning path instructor view

### PR DESCRIPTION
**Pull request by JohannesWt**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I documented the Java code using JavaDoc style.

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

The current instructor view for the learning path feature only displays the learning paths of students in the course. To provide a clearer overview, the instructor view should be redesigned to include more aggregated data.

### Description

This PR introduces a new design for the learning path instructor view. Notifications about whether everything is set up correctly, which were previously displayed at the top of the learning paths table, now have a dedicated section. In this section, the instructor is alerted if competencies or their relationships need to be created, or if some student learning paths have not yet been generated. These notifications are represented by small, color-coded banners indicating the severity of the issue.

Additionally, the learning paths table has been simplified and slightly redesigned. A new analysis section has also been introduced, which provides more aggregated data for the entire course. This section includes a competency path that shows the average mastery of competencies, offering the instructor a clearer overview of student progress.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 2 Students
- No competencies

1. Log in to Artemis
2. Navigate to Instructor Learning path view
3. Verify that the notification "No competencies" is shown
4. Create two or more competencies (without relations)
5. Verify that the notification "No relations" is shown
6. Create relations between the competencies and add a lecture unit to each competency
7. Verify that that the notification "Missing learning paths" is shown
8. Log in as a student and complete the lecture unit
9. Verify as an Instructor that the data in the learning paths table and analytics graph is shown correctly

[](https://artemis-test1.artemis.cit.tum.de)
[](https://artemis-test2.artemis.cit.tum.de)
[](https://artemis-test3.artemis.cit.tum.de)
[](https://artemis-test4.artemis.cit.tum.de)
[](https://artemis-test5.artemis.cit.tum.de)
[](https://artemis-test6.artemis.cit.tum.de)
[](https://artemis-test9.artemis.cit.tum.de)

#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| competency-graph-modal.component.ts | 100% | ✅ |
| competency-node.component.ts | 88.46% | ✅ ❌ |
| learning-path-nav-overview.component.ts | 94.11% | ✅ ❌ |
| learning-paths-analytics.component.ts | 100% | ✅ |
| learning-paths-configuration.component.ts | 94.73% | ✅ ❌ |
| learning-paths-state.component.ts | 100% | ✅ |
| learning-paths-table.component.ts | 97.67% | ✅ ❌ |
| learning-path-management.component.ts | 94.25% | ✅ ❌ |
| learning-path-instructor-page.component.ts | 100% | ✅ |
| base-api-http.service.ts | 95% | ✅ ❌ |
| learning-path-api.service.ts | 95.83% | ✅ ❌ |
| learning-path-health.model.ts | 100% | ✅ |
| learning-path.model.ts | 96.55% | ✅ ❌ |

### Screenshots

<details>
<summary>Before</summary>

image
</details>

<details>
<summary>After</summary>

image
</details>

## Summary by CodeRabbit

- **New Features**
  - Added methods to retrieve non-zero competency progress records and calculate the average mastery progress for specified competencies.
  - Introduced a new method to generate competency graphs for courses.
  - Added a new localization file for German language support in learning path management.
  - Launched a new Angular component for displaying learning paths in a table format with search and pagination functionalities.

- **Bug Fixes**
  - Simplified health status determination in the learning path service by removing the "OK" status.

- **Chores**
  - Updated the constructor of the learning path service to include new repository dependencies.
  - Modified the `LearningPathHealthDTO` to remove the "OK" status from the health status enum.
  - Removed outdated test cases related to health status checks from the LearningPathService tests.

### CODE REVIEW

File: src/test/javascript/spec/component/learning-paths/components/learning-paths-table.component.spec.ts
Comment (pzdr7) [inline src/test/javascript/spec/component/learning-paths/components/learning-paths-table.component.spec.ts:140]: I'll leave it up to you, but as it's currently used, we don't really need the default value for `start`

```suggestion
    function generateResults(start: number, end: number): LearningPathInformationDTO[] {
```

**Review by pzdr7 — CHANGES_REQUESTED** [review]
Code looks good overall 👍 
I have some small comments

File: src/main/webapp/app/course/learning-paths/services/base-api-http.service.ts
Comment (JohannesWt) [reply src/main/webapp/app/course/learning-paths/services/base-api-http.service.ts:231]: Good idea, will da a short follow up PR for this

File: src/main/java/de/tum/cit/aet/artemis/atlas/service/learningpath/LearningPathService.java
Comment (JohannesStoehr) [reply src/main/java/de/tum/cit/aet/artemis/atlas/service/learningpath/LearningPathService.java]: It is now computed in the query.
Regarding the sorting: https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/util/stream/DoubleStream.html#average()
> API Note:
Elements sorted by increasing absolute magnitude tend to yield more accurate results.

**Review by pzdr7 — DISMISSED** [review]
Code

**Review by SimonEntholzer — APPROVED** [review]
New view looks nice on Ts5 :+1:

**Review by az108 — CHANGES_REQUESTED** [review]
I like this new view a lot, good job
I found a little issue where the pagination is not fully working as expected. i have 2000 students in my course and its not giving me the option to switch pages. But i am able to search all students with the search bar

image

**Review by az108 — APPROVED** [review]
works as expected now, ty for the quick fix

**Review by pzdr7 — APPROVED** [review]
Reapprove

**Review by SimonEntholzer — APPROVED** [review]
Reapprove

**Review by JohannesStoehr — APPROVED** [review]
Code

**Review by MaximilianAnzinger — APPROVED** [review]
Maintainer approved

**Pull request merged**
